### PR TITLE
Fix calculation of FPKM

### PR DESCRIPTION
**Pull request by romunov**
This fix addresses the bug which affects calculation of FPKM for results from `featureCounts`. The problem was that raw count file from `featureCounts` comes unsorted and this was a problem because division was made as if sorted. The script now uses merge ("left join") to match data from gtf file and raw count data and does all the calculations.

### CODE REVIEW

File: docs/CHANGELOG.rst [hunk 1]
```diff
@@ -57,6 +57,7 @@ Fixed
   duplicated amplicon names are included. Validation will now pass also
   for primer sequences in lowercase.
 - Fix allele frequency (AF) calculation in ``snpeff`` process
+- Fix bug in script for calculating FPKM
```
Comment (mstajdohar) [inline docs/CHANGELOG.rst]: Provide more info about the bug.

File: resolwe_bio/tools/expression_fpkm_tpm.R [hunk 2]
```diff
@@ -12,31 +13,36 @@ args = parser$parse_args(commandArgs(trailingOnly=TRUE))
...
 annot <- import.gff(args$annotation, format=args$format, genome="NA", feature.type=args$feature_type)
-exon_lengths=width(annot)
-names(exon_lengths)=elementMetadata(annot)[,args$id_attribute]
+exon_lengths <- width(annot)
+names(exon_lengths) <- elementMetadata(annot)[, args$id_attribute]
 
-# calculate gene lengths
-exon_lengths_by_gene=split(exon_lengths, names(exon_lengths))
-gene_lengths=sapply(exon_lengths_by_gene, sum)
+# calculate gene lengths, note that the result is lexographically sorted
```
Comment (mzganec) [inline resolwe_bio/tools/expression_fpkm_tpm.R]: lexicographically

**Comment by mstajdohar** [comment]
How long does the script take for human expressions?

File: docs/CHANGELOG.rst (hunk 1 above)
Comment (mstajdohar) [reply docs/CHANGELOG.rst]: The info should be such that when the CS, product or marketing read about the fix in release notes, they know what it is about. I know you are part of the CS team, but the description should be such that the message does not have to be explained verbally.

Comment (romunov) [reply docs/CHANGELOG.rst]: I'm on it.

File: resolwe_bio/tools/expression_fpkm_tpm.R [hunk 3]
```diff
@@ -12,31 +13,36 @@ args = parser$parse_args(commandArgs(trailingOnly=TRUE))
...
 annot <- import.gff(args$annotation, format=args$format, genome="NA", feature.type=args$feature_type)
-exon_lengths=width(annot)
-names(exon_lengths)=elementMetadata(annot)[,args$id_attribute]
+exon_lengths <- width(annot)
+names(exon_lengths) <- elementMetadata(annot)[, args$id_attribute]
 
-# calculate gene lengths
-exon_lengths_by_gene=split(exon_lengths, names(exon_lengths))
-gene_lengths=sapply(exon_lengths_by_gene, sum)
+# calculate gene lengths, note that the result is lexicographically sorted
```
Comment (JenkoB) [inline resolwe_bio/tools/expression_fpkm_tpm.R]: Change to: # calculate gene or transcript lengths

File: resolwe_bio/tools/expression_fpkm_tpm.R [hunk 4]
```diff
@@ -12,31 +13,36 @@ args = parser$parse_args(commandArgs(trailingOnly=TRUE))
...
+names(ncounts) <- c("Gene", "raw_count")
+ncounts$norm_count <- ncounts$raw_count / sum(ncounts$raw_count) # library_size
 
 # obtain CPM
-cpm=ncounts*1e6
+cpm <- ncounts$norm_count * 1e6
+names(cpm) <- ncounts$Gene
+cpm <- cpm[order(names(cpm))] # lexicographically order cpm
 
 # normalise counts by gene length
```
Comment (JenkoB) [inline resolwe_bio/tools/expression_fpkm_tpm.R]: In the next two lines after this comment you don't normalise by gene length, but you are selecting genes in counts table that you will normalise.

Comment (romunov) [reply resolwe_bio/tools/expression_fpkm_tpm.R]: Thanks, changed.

File: resolwe_bio/tools/expression_fpkm_tpm.R (hunk 3 above)
Comment (romunov) [reply resolwe_bio/tools/expression_fpkm_tpm.R]: Done.

**Pull request merged**